Granular rate limits (for Forms v5)

mbmjertan opened the pull request:
# Description

This PR largely refactors the entire `Security` namespace, and adds support for granular rate limits (per form), in addition to global rate limits.

TL;DR:

* adds a new database table for rate limits
* uses it in the Security class
* adds a post meta to forms for granular rate limiting
* adds a model for log entries
* adds a cleanup service which tries to use Action Scheduler if available or just cleans up if not

Commits should be pretty self-explanatory. Needs heavy testing on our end. We will let you know 😄 

* Closes https://github.com/infobipcth/infobip-web/issues/2954
* See also: https://github.com/infobipcth/eightshift-forms/pull/1
* We will port to v6 as well

mbmjertan commented:
Please take a look at conflict resolution in [8f0297d](https://github.com/infobipcth/eightshift-forms/pull/1/commits/8f0297dac2988467e324186451f862379315ba0b), I'm not sure whether this is okay as I don't fully understand the reasoning behind some changes between 5.6 and 5.9.

iruzevic reviewed (changes requested):
Latest version of forms is 6.1.8 tag https://github.com/infinum/eightshift-forms/releases/tag/6.1.8 please make your patch from that branch if you want this feature to end up in the latest forms release.

If you want to make this feature on the older version of forms (the version of your project) make a feature branch from the tag 5.6.10 https://github.com/infinum/eightshift-forms/releases/tag/5.6.10.

mbmjertan commented:
Ok I will rebase to 5.6.10 and drop your commits

I will later on open a new PR for Forms 6.x once we start work on migrating

mbmjertan commented:
Apart from that do you have concerns regarding this

iruzevic commented:
> Apart from that do you have concerns regarding this

My concerns are:

- will this slow down the requests?
- dont use direct class but use DI or wp filters and actions
- cron job class should go to cronjob events folder not in the security folder
- is the db indexable?
- will the db be created on plugin instalation?


I can give you more details once the proper branches are used as it now contains stuff that was already done by me

mbmjertan commented:
> I can give you more details once the proper branches are used as it now contains stuff that was already done by me
Okay, I'll prepare this tomorrow.
If you want to review in the meantime, this is the commit range I wish to merge: https://github.com/infinum/eightshift-forms/pull/498/files/2a3648120c4593b1df13856f5ec3486756e45ac1..40bf6dd583bb3a4d2c080e51810b6614e9dec6e6

> will this slow down the requests?
To quote Hegel, it depends :D 

Jokes aside, we haven't load tested this nor we have any performance metrics in particular to share. However, what I can comment on is:
* this will decrease memory consumption across routes that perform rate limit validation
  * in the previous implementation, rate limit data was stored as a pre-existing aggregate by IP hashes inside of an option
  * this JSON-serialized nested object could be, depending on the number of requests you store, a big boy (think calculator interactions)
* SQL calls here should perform very quickly, given the indexes defined on the table
  * I've been future-proofing for some more complex logic, so I've added indexes for most pairs a user would query. this is perhaps an unworthwhile trade-off in terms of storage space
*  on our end, we will monitor performance and act accordingly - you can wait for merging until we have some data if you want

> dont use direct class but use DI or wp filters and actions
Not sure I'm following you: where exactly do you mean? The static calls to the "model" object?

> cron job class should go to cronjob events folder not in the security folder
Cool, I noticed this and then I forgot. Thanks, it seems like it has a nice wrapper around Action Scheduler. 🙈 

> is the db indexable?
Yes, I've added indexes. Probably too much.

> will the db be created on plugin instalation?
Yes.

mbmjertan commented:
Done, but I can't change the base branch here and create a new one for the release. See https://github.com/infobipcth/eightshift-forms/pull/1 please

mbmjertan commented:
I will address your comments in new commits soon

mbmjertan commented:
Any updates on this?
